packages/core/core/src/services/document-service/relations/extract-ids/extract-data-ids.ts

```diff
@@ -0,0 +1,105 @@
+import { EntityService, Attribute, Schema } from '@strapi/types';
+import { traverseEntity } from '@strapi/utils';
+import { isObject } from 'lodash/fp';
+
+// TODO: Shorthand id should only be a string on documents.
+// It's also a number here to make it works with existing V4 types.
+type ID = string | number;
+type ShortHand = ID;
+type LongHand = { id: ID };
+
+const isShortHand = (relation: any): relation is string | number => {
+  return typeof relation === 'string' || typeof relation === 'number';
+};
+
+const isLongHand = (relation: any): relation is { id: ID } => {
+  return isObject(relation) && 'id' in relation;
+};
+
+/**
+ * Get relation ids from shorthand id (id)
+ */
+const handleShortHand = (relation: ShortHand): ID[] => [relation];
+
+/**
+ * Get relation ids from arrayed ids (id[], {id}[])
+ */
+const handleArray = (array: ShortHand[] | LongHand[]) =>
+  array.map((item) => (isShortHand(item) ? item : item.id));
+
+/**
+ *  Get relation ids from primitive representation (id, id[], {id}, {id}[])
+ */
+const handlePrimitive = (
+  relation: ShortHand | LongHand | ShortHand[] | LongHand[] | null | undefined
+) => {
+  if (!relation) return []; // null
+  if (isShortHand(relation)) return handleShortHand(relation); // id
+  if (isLongHand(relation)) return handleShortHand(relation.id); // { id }
+  if (Array.isArray(relation)) return handleArray(relation); // id[]
+
+  return [];
+};
+
+/**
+ * Get all relations document ids from a relation input value
+ */
+const extractRelationIdsVisitor = <T extends Attribute.RelationKind.Any>(
+  relation: EntityService.Params.Attribute.RelationInputValue<T>
```

Comment:
@Convly I wanted to share some feedback / difficulties I had while using the relational types!

I don't know if it's a common use case, but I had difficulties splitting this function into other small ones. Because I did not have access to the actual Relational Types, they do not seem to be exposed. And I didn't find a way to narrow this type down.

Also , having to access a relation attribute value type like `EntityService.Params.Attribute...` feels a bit long and difficult to find.

Coming from someone who hasn't worked a lot on the strapi/types package I can only share what would be "ideal" in this scenario, but I don't know if this could introduce other issues:

```
import {Attributes } from "strapi/types"

Attributes.Relation.Format.Connect
Attributes.Relation.Format.Disconnect
Attributes.Relation.Format.Set
Attributes.Relation.Format.ShortHand
Attributes.Relation.Format.LongHand
```

Do you think this is something that would make sense? I wouldn't mind implementing it myself

Reply:
The only main issue that I see with this approach is that the different formats are not related to the attribute itself but to its usage in the entity service (and db layer?).

I don't see how we could justify putting the different format there :thinking: 

However, namespaces do end up being very verbose and we should limit that.

I'm wondering, if -_when using several types from the same nested namespace_- we couldn't just alias the namespace to a regular type and reference other types from there.

For instance:

```ts
import type { EntityService } from '@strapi/types';

type Relations = EntityService.Params.Attribute.Relation // or smth like that

const fn = (x: Relations.LongHand) => { ... }
```

The only issue I see is the loss of context from the created type. But on the other hand, it'll be bound to the file context, so naming shouldn't be too hard.

What do you think?

Reply:
I don't know what I was thinking, indeed that's not possible :thinking: 

I'm not sure what could be the alternative